### PR TITLE
Merging the latest changes from the upstream CAPI project, v0.3.10 included

### PR DESCRIPTION
**Pull request by nprokopic**
What's in the box:
- Merging the latest changes from the upstream CAPI project. This gives us CAPI v0.3.10.
- Update k8s deps to v0.18.6.
- Update controller-runtime to v0.6.3.
    - CAPI tests need `Result.IsZero()` func which was removed in v0.6.0, and was added back in v0.6.2.
    - controller-runtime v0.6.3 is also using k8s v0.18.6 dependency versions.
    
Notes:
- We should keep our fork in sync with the upstream as much as possible, so without some ad-hoc changes.
- The only exception for making changes in our fork should be k8s dependencies.

### CODE REVIEW

File: go.mod
Comment (tuommaki) [inline go.mod:38]: What about those recent bugfixes that went into some `1.18.x` version that affected watching? Probably not a problem in this library, but I was just wondering if we should still align?

Comment (nprokopic) [reply go.mod:38]: I bumped controller-runtime in our fork to v0.6.3, which is using k8s v0.18.6 libraries, so I also bumped k8s dependencies to v0.18.6. See this commit message https://github.com/giantswarm/cluster-api/pull/2/commits/88cd8c1024ff714c49a6ad0d2bc902f2e1698081

Comment (nprokopic) [reply go.mod:38]: TL;DR: needed controller-runtime v0.6.2 for CAPI tests (so I don't have to modify/delete upstream code from our fork), went with v0.6.3 which is the latest. controller-runtime v0.6.3 uses k8s v0.18.6, so I bumped it to that same version for the sake of consistency :)

Comment (tuommaki) [reply go.mod:38]: Ack; All good :+1:

**Review by tuommaki — APPROVED** [review]
LGTM when it compiles :slightly_smiling_face:

**Comment by nprokopic** [comment]
This PR includes upstream changes including v0.3.10, and the latest commit included from the upstream CAPI is the exact one tagged with v0.3.10.

After merging this, we can tag the latest commit in our fork with `v0.3.10-gs` which will make that version from our fork exactly the same as the corresponding upstream version `v0.3.10`, with only difference in k8s dependency versions (and differences that come from that).

**Review by rossf7 — APPROVED** [review]
LGTM 🚀

**Comment by nprokopic** [comment]
It compiles, but CAPI unit tests are failing, but that's also the case on the master branch. I guess it's not an issue for our use case.

**Comment by tuommaki** [comment]
> I guess it's not an issue for our use case.

Nope. If I remember correctly there was some problem with the `kubebuilder` test env setup or something. Not serious since we don't run their controllers.

**Pull request merged**
